### PR TITLE
Introduce Arr_observer<Arrangement_2>

### LINKED ISSUE #7901: Pointer Authentification issues in Arrangement
On modern Apple architectures, [pointer authentification](https://developer.apple.com/documentation/security/preparing_your_app_to_work_with_pointer_authentication) is used to detect if a pointer is used as a given type but was actually created from another type.

For instance, the following example produces such a runtime error:
https://doc.cgal.org/latest/Arrangement_on_surface_2/Arrangement_on_surface_2_2observer_8cpp-example.html

It comes from the fact that the arrangement type in the base observer is not the real type that will be used later internally.

Doing the following changes:
https://github.com/CGAL/cgal/compare/master...sloriot:cgal:argt-PAC?expand=1
where I basically add a final type for the arrangement in the base class makes the error to go away.

One alternative (cheapest) solution would be to replace Arrangement_2 class by a template alias to Arrangement_on_surface_2, probably adding some extra methods in AoS_2 class.

What do you think @efifogel ?

### PR DESCRIPTION
**Pull request by efifogel**
## Summary of Changes

1. Renamed the template Arr_observer to Aos_observer
2. Added the nested type 'Observer' in the Aos_2 class template, which is an alias to Arr_observer<Aos_2>.
3. Finally, introduced Arr_observer<Arrangement_2>, which is an alias to Arr_observer<Aos_2>, where Arrangement_2 derives from Aos_2 (for cases where Arrangement_2 is not Aos_2).

The first 2 items eliminate the error-prone c-type casting that was used to define observers.
The 3rd item is for backward compatibility.

## Release Management

* Affected package(s): Arrangement_on_surface_2, Minkowski_sum_2, Visibility_2, and Envelope_3
* Issue(s) solved (if any): fix #7901
* Feature/Small Feature (if any):
* Link to compiled documentation (obligatory for small feature) [*wrong link name to be changed*](httpssss://wrong_URL_to_be_changed/Manual/Pkg)
* License and copyright ownership:

### CODE REVIEW

File: Arrangement_on_surface_2/doc/Arrangement_on_surface_2/CGAL/Aos_observer.h [hunk 1]
```diff
@@ -0,0 +1,409 @@
+namespace CGAL {
+
+/*! \ingroup PkgArrangementOnSurface2Ref
+ *
+ * \anchor arr_refaos_obs
+ *
+ * `Aos_observer` serves as an abstract base class for all observer classes that
+ * are attached to an arrangement instance of type `Arrangement` and receive
+ * notifications from the arrangement.  This base class handles the attachment
```
Comment (albert-github) [inline Arrangement_on_surface_2/doc/Arrangement_on_surface_2/CGAL/Aos_observer.h:9]: 2 spaces should be 1 space (also on other places)

File: Arrangement_on_surface_2/doc/Arrangement_on_surface_2/CGAL/Aos_observer.h [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+  /*! the type of the associated arrangement. */
+  typedef unspecified_type Arrangement_2;
+
+  /*! the point type. */
+  typedef typename Arrangement_2::Point_2 Point_2;
+
+  /*! the \f$ x\f$-monotone curve type. */
+  typedef typename Arrangement_2::X_monotone_curve_2 X_monotone_curve_2;
+
+  /*! */
```
Comment (albert-github) [inline Arrangement_on_surface_2/doc/Arrangement_on_surface_2/CGAL/Aos_observer.h:39]: To be filled in? (also on the next few lines?)

File: Arrangement_on_surface_2/doc/Arrangement_on_surface_2/CGAL/Aos_observer.h [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+  virtual void after_create_boundary_vertex(Vertex_handle v);
+
+  /*! issued just before a new edge that corresponds to the \f$ x\f$-monotone
+   * curve `c` and connects the vertices `v1` and `v2` is created.
+   */
+  virtual void before_create_edge(const X_monotone_curve_2& c,
+                                  Vertex_handle v1, Vertex_handle v2);
+
+  /*! issued immediately after a new edge `e` has been created.  The halfedge that
+   * is sent to this function is always directed from `v1` to `v2` (see above).
```
Comment (albert-github) [inline Arrangement_on_surface_2/doc/Arrangement_on_surface_2/CGAL/Aos_observer.h:172]: Better refer to the function name as e.g. due to sorting the order in the documentation can change.

File: Arrangement_on_surface_2/include/CGAL/Aos_observer.h [hunk 4]
```diff
@@ -0,0 +1,573 @@
...
+#define CGAL_AOS_OBSERVER_H
+
+#include <CGAL/license/Arrangement_on_surface_2.h>
+
+#include <CGAL/disable_warnings.h>
+
+#include <CGAL/Arr_enums.h>
+
+/*! \file
+ * Definition of the Aos_observer<Arrangement> base class.
```
Comment (albert-github) [inline Arrangement_on_surface_2/include/CGAL/Aos_observer.h:24]: backticks around `Aos_observer<Arrangement>`?

File: Arrangement_on_surface_2/include/CGAL/Aos_observer.h [hunk 5]
```diff
@@ -0,0 +1,573 @@
...
+    // Unregister the observer object from the arrangement.
+    if (p_arr != nullptr)
+      p_arr->_unregister_observer(this);
+  }
+  //@}
+
+  /// \name Modifying the associated arrangement.
+  //@{
+
+  /*! Get the associated arrangement (non-const version). */
```
Comment (albert-github) [inline Arrangement_on_surface_2/include/CGAL/Aos_observer.h:84]: "non-const" version?

File: Arrangement_on_surface_2/include/CGAL/Aos_observer.h [hunk 6]
```diff
@@ -0,0 +1,573 @@
...
+
+  /*! Notification after a global operation is completed. */
+  virtual void after_global_change() {}
+  //@}
+
+  /// \name Notification functions on observer attachment or detachment.
+  //@{
+
+  /*! Notification before the observer is attached to an arrangement.
+   * \param arr The arrangement we are about to attach the observer to.
```
Comment (albert-github) [inline Arrangement_on_surface_2/include/CGAL/Aos_observer.h:163]: is the sentence complete? (probably text "an arrangement" is missing)

File: Arrangement_on_surface_2/include/CGAL/Aos_observer.h [hunk 7]
```diff
@@ -0,0 +1,573 @@
...
+   */
+  virtual void before_attach(const Arrangement_2& /* arr */) {}
+
+  /*! Notification after the observer has been attached to an arrangement. */
+  virtual void after_attach() {}
+
+  /*! Notification before the observer is detached from the arrangement. */
+  virtual void before_detach() {}
+
+  /*! Notification after the observer has been detached to the arrangement. */
```
Comment (albert-github) [inline Arrangement_on_surface_2/include/CGAL/Aos_observer.h:173]: "to the arrangement" -> "from the arrangement"

File: Arrangement_on_surface_2/include/CGAL/Aos_observer.h [hunk 8]
```diff
@@ -0,0 +1,573 @@
...
+  virtual void after_detach() {}
+  //@}
+
+  /// \name Notification functions on local changes in the arrangement.
+  //@{
+
+  /*!
+   * Notification before the creation of a new vertex.
+   * \param p The point to be associated with the vertex.
+   *          This point cannot lies on the surface boundaries.
```
Comment (albert-github) [inline Arrangement_on_surface_2/include/CGAL/Aos_observer.h:183]: "lies" -> "lie"

File: Arrangement_on_surface_2/include/CGAL/Aos_observer.h [hunk 9]
```diff
@@ -0,0 +1,573 @@
...
+  virtual void before_create_vertex(const Point_2& /* p */) {}
+
+  /*! Notification after the creation of a new vertex.
+   * \param v A handle to the created vertex.
+   */
+  virtual void after_create_vertex(Vertex_handle /* v */)
+  {}
+
+  /*! Notification before the creation of a new boundary vertex.
+   * \param p The on the surface boundary.
```
Comment (albert-github) [inline Arrangement_on_surface_2/include/CGAL/Aos_observer.h:194]: "The on" ?

File: Arrangement_on_surface_2/include/CGAL/Aos_observer.h [hunk 10]
```diff
@@ -0,0 +1,573 @@
...
+                                             Arr_parameter_space /* ps_y */)
+  {}
+
+  /*! Notification after the creation of a new vertex at infinity.
+   * \param v A handle to the created vertex.
+   */
+  virtual void after_create_boundary_vertex(Vertex_handle /* v */) {}
+
+  /*! Notification before the creation of a new edge.
+   * \param c The x-monotone curve to be associated with the edge.
```
Comment (albert-github) [inline Arrangement_on_surface_2/include/CGAL/Aos_observer.h:221]: `x-monotone` elsewhere it is like a formula `\f$ x\f$-monotone` (also at other places)

File: Arrangement_on_surface_2/include/CGAL/Aos_observer.h [hunk 11]
```diff
@@ -0,0 +1,573 @@
...
+   * \param v1 A handle to the first end-vertex of the edge.
+   * \param v2 A handle to the second end-vertex of the edge.
```
Comment (albert-github) [inline Arrangement_on_surface_2/include/CGAL/Aos_observer.h:223]: 2 end-vertices?

File: Arrangement_on_surface_2/doc/Arrangement_on_surface_2/CGAL/Aos_observer.h [hunk 12]
```diff
@@ -0,0 +1,409 @@
...
+  virtual void before_split_face(Face_handle f, Halfedge_handle e);
+
+  /*! issued immediately after the existing face `f1` has been split, such that a
+   * portion of it now forms a new face `f2`. The flag `is_hole` designates
+   * whether `f2` forms a hole (an inner CCB) inside `f1`.
+   */
+  virtual void after_split_face(Face_handle f1, Face_handle f2, bool is_hole);
+
+  /*! issued just before outer ccb `h` inside a face `f` is split into two, as a
+   * result of the removal of the edge `e` from the arrangement.
```
Comment (albert-github) [inline Arrangement_on_surface_2/doc/Arrangement_on_surface_2/CGAL/Aos_observer.h:229]: `ccb` or `CCB` (just above and in another file I saw `CCB`)

**Comment by afabri** [comment]
Error in the [documentation](https://cgal.geometryfactory.com/CGAL/Manual_doxygen_test/CGAL-6.0-Ic-127/) : `CGAL-6.0-Ic-127/doc/scripts/doc_1_8_13/doc_tags/Arrangement_on_surface_2.tag:11511: warning: Duplicate anchor arr_refarr_obs found`

**Comment by efifogel** [comment]
All fixed. Thanks.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by efifogel** [comment]
It is defined in one place only, so I don't understand how or why this
warning message is generated
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by albert-github** [comment]
Regarding the documentation warning when running doxygen 1.8.13
- this might be a bug in the doxygen 1.8.13 version
- is thus pull request just for the 6.0 and newer versions or will it also be in the 5.5 / 5.6 CGAL versions
   - in case it is only intended for the 6.0 and newer versions there won't be a problem as the doxygen 1.9.6 version will be used here (to the best of my knowledge)
   - in case it is also back-ported to 5.5 and 5.6 we have to see what the effect on the documentation is

**Comment by efifogel** [comment]
Indeed, when running locally using doxygen version 1.9.7 the warning is
*not* generated.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by sloriot** [comment]
New warnings:
```
/mnt/testsuite/include/CGAL/Arr_trapezoid_ric_point_location.h:241:16: warning: 'CGAL::Arr_trapezoid_ric_point_location<CGAL::Arrangement_2<CGAL::Arr_segment_traits_2<>>>::before_assign' hides overloaded virtual function [-Woverloaded-virtual]
  241 |   virtual void before_assign (const Arrangement_on_surface_2& arr)
      |                ^
/mnt/testsuite/test/Arrangement_on_surface_2/test_vert_ray_shoot_vert_segments.cpp:60:10: note: in instantiation of template class 'CGAL::Arr_trapezoid_ric_point_location<CGAL::Arrangement_2<CGAL::Arr_segment_traits_2<>>>' requested here
   60 |   RIC_pl ric_pl(arr);
      |          ^
/mnt/testsuite/include/CGAL/Aos_observer.h:139:16: note: hidden overloaded virtual function 'CGAL::Aos_observer<CGAL::Arrangement_on_surface_2<CGAL::Arr_segment_traits_2<>, CGAL::Arr_bounded_planar_topology_traits_2<CGAL::Arr_segment_traits_2<>>>>::before_assign' declared here: type mismatch at 1st parameter ('const Arrangement_2 &' (aka 'const CGAL::Arrangement_on_surface_2<CGAL::Arr_segment_traits_2<>, CGAL::Arr_bounded_planar_topology_traits_2<CGAL::Arr_segment_traits_2<>>> &') vs 'const Arrangement_on_surface_2 &' (aka 'const CGAL::Arrangement_2<CGAL::Arr_segment_traits_2<>> &'))
  139 |   virtual void before_assign(const Arrangement_2& /* arr */) {}
      |                ^
In file included from /mnt/testsuite/test/Arrangement_on_surface_2/test_vert_ray_shoot_vert_segments.cpp:6:
/mnt/testsuite/include/CGAL/Arr_trapezoid_ric_point_location.h:263:16: warning: 'CGAL::Arr_trapezoid_ric_point_location<CGAL::Arrangement_2<CGAL::Arr_segment_traits_2<>>>::before_attach' hides overloaded virtual function [-Woverloaded-virtual]
  263 |   virtual void before_attach (const Arrangement_on_surface_2& arr)
```

**Comment by sloriot** [comment]
Might also be responsible for runtime errors here:
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.0-Ic-127/Surface_sweep_2/TestReport_lrineau_Ubuntu-Latest.gz

**Comment by efifogel** [comment]
clang is good (it was clang that detected this problem, right?)
This was a real problem.
The basic idea behind this entire fix is that everything that has to do
with the observer is based on the very base Arrangement_on_surface  type.
For example, if we have A1 derived from A2 derived ... derived from
CGAL::Arrangement_on_surface_2, the observer is based on the latter, and so
are the types of the parameters of the virtual functions that accept an
arrangement type.)
Anyway, I fixed this last problem; please try again.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by efifogel** [comment]
The run time error does not seem related, but who knows... let's try one
thing at a time.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
> New warnings:
> 
> ```
> /mnt/testsuite/include/CGAL/Arr_trapezoid_ric_point_location.h:241:16: warning: 'CGAL::Arr_trapezoid_ric_point_location<CGAL::Arrangement_2<CGAL::Arr_segment_traits_2<>>>::before_assign' hides overloaded virtual function [-Woverloaded-virtual]
>   241 |   virtual void before_assign (const Arrangement_on_surface_2& arr)
>       |                ^

> ```

Now that we use C++11 or later, prefer the use of `override` instead of `virtual`. That way all compilers will emit an error if a member function supposed to override a virtual function of a base class have a mismatch in the parameters. The specifier `virtual` is only for the base function of the hierarchy.

**Comment by sloriot** [comment]
New warnings:

- [ ] `/Users/magritte/cgal_root/CGAL-6.0-Ic-131/include/CGAL/Arr_trapezoid_ric_point_location.h:301:16: warning: 'CGAL::Arr_trapezoid_ric_point_location<CGAL::Arrangement_with_history_2<CGAL::Arr_segment_traits_2<>>>::before_split_edge' hides overloaded virtual function [-Woverloaded-virtual]   virtual void before_split_edge (Halfedge_handle e,`
               
- [ ] `/Users/magritte/cgal_root/CGAL-6.0-Ic-131/include/CGAL/Arr_trapezoid_ric_point_location.h:325:16: warning: 'CGAL::Arr_trapezoid_ric_point_location<CGAL::Arrangement_with_history_2<CGAL::Arr_segment_traits_2<>>>::before_merge_edge' hides overloaded virtual function [-Woverloaded-virtual]  virtual void before_merge_edge (Halfedge_handle e1,`

**Comment by efifogel** [comment]
Right.
It doesn't end there.
I went over all observers and fixed the definition of the ised types within
to use the base arrangement.
All tests pass, but more changes might be needed.
Anyhow, the code is ready.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by efifogel** [comment]
Yep.
I added the override specifier to all virtual functions of the observer.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by sloriot** [comment]
New warnings:
- `/Users/magritte/cgal_root/CGAL-6.0-Ic-137/test/Arrangement_on_surface_2/test_observer.cpp:406:16: warning: 'after_merge_outer_ccb' overrides a member function but is not marked 'override' [-Winconsistent-missing-override]`
- `/Users/magritte/cgal_root/CGAL-6.0-Ic-137/include/CGAL/Aos_observer.h:451:16: note: overridden virtual function is here
  virtual void after_merge_outer_ccb(Face_handle /* f */,`

**Comment by efifogel** [comment]
Yep; forgot one override.
I fixed it.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by sloriot** [comment]
Tests are looking OK in CGAL-6.0-Ic-156 but we are still having a segfault on Arm64e [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.0-Ic-156/Arrangement_on_surface_2/TestReport_Taylor_macOS-Apple_M2-boost_backend-arm64e.gz)

I should have a look locally.

**Comment by sloriot** [comment]
The issue is not specific to macOs but requires to disable GMP. The branch can be merged I'll handle the issue in another branch.

**Comment by efifogel** [comment]
Any hint for reproduction on a non-arm machine?
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by sloriot** [comment]
The issue comes AFAIU from the reading of input into `Quotient<MP_Float>`. If you configure the test with `-DCGAL_DISABLE_GMP=ON` you should be able to reproduce it on your machine.

**Comment by efifogel** [comment]
oh, so it seems unrelated to the arr. observer.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by sloriot** [comment]
It's a bug in `Quotient.h`, I'll fix it.

**Pull request merged**